src/librustc_mir/borrow_check/nll/region_infer/error_reporting/outlives_suggestion.rs

```diff
@@ -0,0 +1,254 @@
+    fn region_vid_to_name(
+        &self,
+        errctx: &ErrorReportingCtx<'_, '_, '_>,
+        renctx: &mut RegionErrorNamingCtx,
+        region: RegionVid,
+    ) -> Option<RegionName> {
+        errctx
+            .rinfcx
+            .give_region_a_name(errctx, renctx, region)
+            .filter(Self::region_name_is_suggestable)
+    }
+
+    /// Add the outlives constraint `fr: outlived_fr` to the set of constraints we need to suggest.
+    crate fn collect_constraint(&mut self, fr: RegionVid, outlived_fr: RegionVid) {
+        debug!("Collected {:?}: {:?}", fr, outlived_fr);
+
+        // Add to set of constraints for final help note.
+        if let Some(ref mut outlived_frs) = self.constraints_to_add.get_mut(&fr) {
+            outlived_frs.push(outlived_fr);
+        } else {
+            self.constraints_to_add.insert(fr, vec![outlived_fr]);
+        }
+    }
+
+    /// Emit an intermediate note on the given `Diagnostic` if the involved regions are
+    /// suggestable.
+    crate fn intermediate_suggestion(
+        &mut self,
+        errctx: &ErrorReportingCtx<'_, '_, '_>,
+        errci: &ErrorConstraintInfo,
+        renctx: &mut RegionErrorNamingCtx,
+        diag: &mut DiagnosticBuilder<'_>,
+    ) {
+        // Emit an intermediate note.
+        let fr_name = self.region_vid_to_name(errctx, renctx, errci.fr);
+        let outlived_fr_name = self.region_vid_to_name(errctx, renctx, errci.outlived_fr);
+
+        match (fr_name, outlived_fr_name) {
+            (Some(fr_name), Some(outlived_fr_name)) => {
+                if let RegionNameSource::Static = outlived_fr_name.source {
+                    diag.help(&format!(
+                        "Consider replacing `{}` with `'static`", fr_name
+                    ));
+                } else {
+                    diag.help(&format!(
+                        "Consider adding the following bound: `{}: {}`",
+                        fr_name, outlived_fr_name
+                    ));
+                }
+            }
+
+            // Not both suggestable
+            _ => {}
+        }
+    }
+
+    /// If there is a suggestion to emit, add a diagnostic to the buffer. This is the final
+    /// suggestion including all collected constraints.
+    crate fn add_suggestion<'tcx>(
+        &self,
+        body: &Body<'tcx>,
+        rinfcx: &RegionInferenceContext<'tcx>,
+        infcx: &InferCtxt<'_, 'tcx>,
+        errors_buffer: &mut Vec<Diagnostic>,
+        renctx: &mut RegionErrorNamingCtx,
+    ) {
+        // No constraints to add? Done.
+        if self.constraints_to_add.is_empty() {
+            debug!("No constraints to suggest.");
+            return;
+        }
+
+        // Create a new diagnostic.
+        let mut diag = DiagnosticBuilder::new(
+            infcx.tcx.sess.diagnostic(),
+            Level::Help,
+            "the following changes may resolve your lifetime errors",
+        );
```

Comment:
What's the benefit of having the individual help without a span telling the user the same thing as the sub-help in the lifetime error?

Reply:
@estebank This message collects all of the suggestions together for the entire MIR item, rather than just a single lifetime error. This means we can give better suggestions here than the individual messages.

For example, if two sub-help messages for different errors may say `add 'a: 'b` and `add 'b: 'a`, but the overall message would say `consider unifying 'a and 'b, since they must be the same`.

One option is to just filter out the overall suggestion if there is only one suggested lifetime bound. I think this could get rid of some redundancy.

Reply:
Implemented in most recent commits.